Prose/ProseLib/Tests/ProseCoreTCATests/TypesEncodingTests.swift

```diff
@@ -0,0 +1,106 @@
+//
+// This file is part of prose-app-macos.
+// Copyright (c) 2022 Prose Foundation
+//
+
+import ProseCoreTCA
+import XCTest
+
+final class TypesEncodingTests: XCTestCase {
+  let encoder: JSONEncoder = {
+    let encoder = JSONEncoder()
+    // Sort object keys so results are predictable
+    encoder.outputFormatting = .sortedKeys
+    return encoder
+  }()
+
+  func check<T: Encodable>(_ value: T, equals expected: String) throws {
+    let data = try self.encoder.encode(value)
+    let string = String(data: data, encoding: .utf8).expect("\(value) has invalid UTF-8")
+    XCTAssertEqual(string, expected)
+  }
+
+  func check<T: Encodable>(_ value: T, in expectedOneOf: [String]) throws {
+    let data = try self.encoder.encode(value)
+    let string = String(data: data, encoding: .utf8).expect("\(value) has invalid UTF-8")
+    XCTAssertTrue(expectedOneOf.contains(string))
+  }
+
+  func testJIDEncoding() throws {
+    func test(_ jid: JID, equals expected: String) throws {
+      try self.check(jid, equals: expected)
+    }
+
+    try test("test@prose.org", equals: #""test@prose.org""#)
+    try test("test@prose.org/resource", equals: #""test@prose.org""#)
+  }
+
+  func testReactionEncoding() throws {
+    func test(_ reaction: Reaction, equals expected: String) throws {
+      try self.check(reaction, equals: expected)
+    }
+
+    try test("😀", equals: #""😀""#)
+    try test("test", equals: #""test""#)
+  }
+
+  func testMessageReactionsEncoding() throws {
+    func test(_ reactions: MessageReactions, equals expected: String) throws {
+      try self.check(reactions, equals: expected)
+    }
+    func test(_ reactions: MessageReactions, in expectedOneOf: [String]) throws {
+      try self.check(reactions, in: expectedOneOf)
+    }
+
+    try test(
+      ["😀": ["tests1@prose.org"], "🧪": ["tests2@prose.org"]],
+      equals: #"[{"authors":["tests1@prose.org"],"reaction":"😀"},{"authors":["tests2@prose.org"],"reaction":"🧪"}]"#
+    )
+    try test(
+      ["🧪": ["tests2@prose.org"], "😀": ["tests1@prose.org"]],
+      equals: #"[{"authors":["tests2@prose.org"],"reaction":"🧪"},{"authors":["tests1@prose.org"],"reaction":"😀"}]"#
+    )
+    try test(
+      ["🧪": [], "😀": ["tests1@prose.org"]],
+      equals: #"[{"authors":["tests1@prose.org"],"reaction":"😀"}]"#
+    )
+    // Reaction authors is stored in a `Set`, so we can't predict the output
+    try test(
+      ["2️⃣": ["tests1@prose.org", "tests2@prose.org"]],
+      in: [
+        #"[{"authors":["tests1@prose.org","tests2@prose.org"],"reaction":"2️⃣"}]"#,
+        #"[{"authors":["tests2@prose.org","tests1@prose.org"],"reaction":"2️⃣"}]"#,
```

Comment:
Interesting approach. We could also sort the authors when encoding. That would probably have a negligible performance impact since it only affects reactions with multiple authors - although that might depend on the size of the workspace.

Another approach would be to pass the view only what it needs, which is the emoji, a count and a flag if one of those reactions is mine (current user).

Reply:
Or, as I was doing a few days ago, use an `OrderedSet` 🙃 For the same reason as in my next sentence, We will go back to an `OrderedSet` one day or another 😌

> Another approach would be to pass the view only what it needs

At some point we will show this data in the view, so it's better like this

Reply:
> We could also sort the authors when encoding

I thought about it, but it's only useful for the tests, and you didn't want me to use `OrderedSet`, so I believed it wasn't a good idea

Reply:
Ok, I put that here for the record. I completely forgot about the feature where you could see the author of each reaction when hovering over the reaction tag. And after a quick test with you on Slack these seem to be ordered by timestamp. So you were right the first time to use a OrderedSet and we should reintroduce that.